File: benchmark/ruby/heap_sort.rb
start = Time.now.to_f * 1000
arr = (1..1000).to_a.reverse
def heapify(arr, n, idx)
    largest = idx
    left = 2 * idx + 1
    right = 2 * idx + 2
    largest = left if left < n && arr[left] > arr[largest]
    largest = right if right < n && arr[right] > arr[largest]
    if largest != idx
        arr[idx], arr[largest] = arr[largest], arr[idx]
        heapify(arr, n, largest)
    end
end
n = arr.length
(n / 2 - 1).downto(0) { |j| heapify(arr, n, j) }
(n - 1).downto(1) do |k|
    arr[0], arr[k] = arr[k], arr[0]
    heapify(arr, k, 0)
end
finish = Time.now.to_f * 1000
puts "First: #{arr[0]}"
puts "Last: #{arr[-1]}"
puts "Time: #{(finish - start).round}ms"

File: benchmark/perl/heap_sort.pl
use Time::HiRes qw(time);
my $n = 1000;
my @arr = reverse 1..$n;
my $start = time();
for (my $i = int($n/2) - 1; $i >= 0; $i--) {
    my $idx = $i;
    while (1) {
        my $largest = $idx; my $left = 2*$idx+1; my $right = 2*$idx+2;
        if ($left < $n && $arr[$left] > $arr[$largest]) { $largest = $left; }
        if ($right < $n && $arr[$right] > $arr[$largest]) { $largest = $right; }
        last if $largest == $idx;
        @arr[$idx, $largest] = @arr[$largest, $idx];
        $idx = $largest;
    }
}
for (my $i = $n - 1; $i > 0; $i--) {
    @arr[0, $i] = @arr[$i, 0];
    my $idx = 0; my $size = $i;
    while (1) {
        my $largest = $idx; my $left = 2*$idx+1; my $right = 2*$idx+2;
        if ($left < $size && $arr[$left] > $arr[$largest]) { $largest = $left; }
        if ($right < $size && $arr[$right] > $arr[$largest]) { $largest = $right; }
        last if $largest == $idx;
        @arr[$idx, $largest] = @arr[$largest, $idx];
        $idx = $largest;
    }
}
my $end = time();
my $ms = ($end - $start) * 1000;
print "First: $arr[0]
";
print "Last: $arr[$n-1]
";
printf "Time: %.0fms
", $ms;
